File: instabiz/instabiz/page/ib_bank_statement_import/ib_bank_statement_import.py
"""IB Bank Statement Import — generic CSV column-mapper (any bank), creates Bank
Transaction records, and auto-creates + reconciles a Payment Entry against any
transaction whose amount confidently matches exactly one open Sales/Purchase
document. Everything else is left as an unreconciled Bank Transaction for the
native Bank Reconciliation Tool.
"""
import csv
import io
import json
import re
from datetime import datetime

import frappe
from frappe import _
from frappe.utils import flt

from erpnext.accounts.doctype.bank_reconciliation_tool.bank_reconciliation_tool import (
	create_payment_entry_bts,
	reconcile_vouchers,
)

from instabiz.overrides.billing_mode import (
	purchase_doctype,
	purchase_outstanding_expr,
	sales_doctype,
	sales_outstanding_expr,
)


def _parse_amount(val):
	"""Remove Indian-format commas and return float."""
	if not val:
		return 0.0
	return flt(re.sub(r",", "", str(val).strip()))


def _parse_date(val):
	"""DD/MM/YYYY → YYYY-MM-DD (also accepts a few other common export formats)."""
	val = str(val).strip()
	for fmt in ("%d/%m/%Y", "%d-%m-%Y", "%Y-%m-%d", "%d/%m/%y", "%m/%d/%Y"):
		try:
			return datetime.strptime(val, fmt).strftime("%Y-%m-%d")
		except ValueError:
			continue
	frappe.throw(_("Cannot parse date: {0}").format(val))


def _find_header_row(lines):
	"""Return index of the CSV header row — the first row containing at least
	two non-empty cells, used as a best-effort skip past bank letterhead/summary
	junk rows that precede the real column headers in most bank exports."""
	for i, line in enumerate(lines):
		cells = [c.strip() for c in line.split(",") if c.strip()]
		if len(cells) >= 3:
			return i
	frappe.throw(_("Could not find a header row in this file."))


@frappe.whitelist()
def get_csv_headers(csv_text):
	"""Return raw column headers + a few sample rows, for the column-mapping UI."""
	lines = csv_text.splitlines()
	hdr_idx = _find_header_row(lines)
	reader = csv.DictReader(io.StringIO("\n".join(lines[hdr_idx:])))
	headers = reader.fieldnames or []
	samples = []
	for row in reader:
		samples.append(row)
		if len(samples) >= 3:
			break
	return {"headers": headers, "samples": samples}


@frappe.whitelist()
def get_saved_profile(bank_account):
	"""Return this bank account's saved column-mapping profile, if any."""
	frappe.has_permission("IB Bank Import Profile", "read", throw=True)
	if not frappe.db.exists("IB Bank Import Profile", bank_account):
		return None
	return frappe.get_doc("IB Bank Import Profile", bank_account).as_dict()


@frappe.whitelist()
def save_profile(bank_account, profile):
	"""Create/update the column-mapping profile for this bank account, so future
	imports for the same bank skip the mapping step entirely."""
	frappe.has_permission("IB Bank Import Profile", "write", throw=True)
	if isinstance(profile, str):
		profile = frappe.parse_json(profile)

	fields = ["mode", "col_date", "col_description", "col_reference",
	          "col_debit", "col_credit", "col_amount", "col_type", "credit_indicator"]
	if frappe.db.exists("IB Bank Import Profile", bank_account):
		doc = frappe.get_doc("IB Bank Import Profile", bank_account)
	else:
		doc = frappe.new_doc("IB Bank Import Profile")
		doc.bank_account = bank_account

	for f in fields:
		doc.set(f, profile.get(f))
	doc.save(ignore_permissions=True)
	frappe.db.commit()
	return {"name": doc.name}


def _parse_csv_with_profile(csv_text, profile):
	"""Parse CSV using an explicit column mapping (no per-bank guessing).

	profile: dict with mode ("Separate Debit/Credit Columns" | "Single Amount
	Column"), col_date, col_description, col_reference, and either
	(col_debit, col_credit) or (col_amount, col_type, credit_indicator).
	"""
	lines = csv_text.splitlines()
	hdr_idx = _find_header_row(lines)
	reader = csv.DictReader(io.StringIO("\n".join(lines[hdr_idx:])))

	mode = profile.get("mode")
	col_date = profile.get("col_date")
	col_desc = profile.get("col_description")
	col_ref = profile.get("col_reference")
	col_debit = profile.get("col_debit")
	col_credit = profile.get("col_credit")
	col_amount = profile.get("col_amount")
	col_type = profile.get("col_type")
	credit_indicator = (profile.get("credit_indicator") or "").strip().lower()

	rows = []
	for row in reader:
		raw_date = (row.get(col_date) or "").strip()
		if not raw_date:
			continue
		try:
			date = _parse_date(raw_date)
		except Exception:
			continue

		description = (row.get(col_desc) or "").strip() if col_desc else ""
		reference_no = (row.get(col_ref) or "").strip() if col_ref else ""

		if mode == "Single Amount Column":
			amt = _parse_amount(row.get(col_amount))
			if col_type:
				is_credit = (row.get(col_type) or "").strip().lower() == credit_indicator
				withdrawal = 0.0 if is_credit else abs(amt)
				deposit = abs(amt) if is_credit else 0.0
			else:
				# Signed amount convention: positive = credit/deposit, negative = debit/withdrawal.
				withdrawal = abs(amt) if amt < 0 else 0.0
				deposit = amt if amt > 0 else 0.0
		else:
			withdrawal = _parse_amount(row.get(col_debit)) if col_debit else 0.0
			deposit = _parse_amount(row.get(col_credit)) if col_credit else 0.0

		if withdrawal == 0 and deposit == 0:
			continue

		rows.append({
			"date":             date,
			"description":      description,
			"reference_number": reference_no,
			"withdrawal":       withdrawal,
			"deposit":          deposit,
		})

	return rows


@frappe.whitelist()
def preview_statement(bank_account, csv_text, profile):
	"""Return parsed rows for client-side preview table."""
	frappe.has_permission("Bank Transaction", "create", throw=True)
	if isinstance(profile, str):
		profile = frappe.parse_json(profile)
	rows = _parse_csv_with_profile(csv_text, profile)
	return {
		"rows":          rows,
		"count":         len(rows),
		"total_deposit": sum(r["deposit"] for r in rows),
		"total_withdrawal": sum(r["withdrawal"] for r in rows),
	}


def _find_confident_match(amount, receiving, company=None):
	"""Exactly one open Sales/Purchase document with this outstanding amount,
	system-wide (scoped to the bank's own company, if given) — anything less
	certain (0 or 2+ candidates) is left for manual reconciliation via the
	native Bank Reconciliation Tool. Respects instabiz.overrides.billing_mode
	so this matches against Sales/Purchase Order in dev mode (billing isn't
	live) the same way the rest of the app does.

	Company filter: `import_statement()` already resolves the bank account's
	own company but never passed it down here, so a deposit could in principle
	auto-match an open order/invoice belonging to a different company (single-
	company today, so currently a no-op in practice — kept as a real guard,
	not dead code, since this app's own per-location accounting has been
	moving toward stricter separation, see the per-location stock account
	work). Left unscoped (None) only preserves the prior no-filter behavior
	for any caller that doesn't have a company to hand.
	"""
	if receiving:
		dt, expr, party_field, party_type = sales_doctype(), sales_outstanding_expr("t"), "customer", "Customer"
	else:
		dt, expr, party_field, party_type = purchase_doctype(), purchase_outstanding_expr("t"), "supplier", "Supplier"

	company_cond = "AND t.company = %(company)s" if company else ""
	rows = frappe.db.sql(
		f"""
		SELECT t.name, t.{party_field} AS party
		FROM `tab{dt}` t
		WHERE t.docstatus = 1 AND ROUND({expr}, 2) = %(amount)s
		{company_cond}
		""",
		{"amount": round(flt(amount), 2), "company": company},
		as_dict=True,
	)
	if len(rows) == 1 and rows[0].party:
		return rows[0].name, rows[0].party, party_type, dt
	return None, None, None, None


def _try_auto_match(bt_name, date, company=None):
	"""Best-effort — never blocks the import if matching/PE-creation fails.

	Builds the Payment Entry via ERPNext's own create_payment_entry_bts (proven
	account/currency/exchange-rate resolution) with allow_edit=True so it's
	returned unsaved, then attaches an explicit, fully-allocated reference to
	the one specific document that matched — rather than relying on
	instabiz.overrides.payment_entry._auto_reconcile's generic FIFO fallback,
	which only handles Receive/Customer against Sales Invoice and would miss
	both Sales-Order-based dev-mode billing and the entire Purchase/Pay side.
	"""
	try:
		bt = frappe.db.get_value(
			"Bank Transaction", bt_name, ["deposit", "withdrawal", "reference_number"], as_dict=True
		)
		amount = bt.deposit or bt.withdrawal
		if not amount:
			return False

		doc_name, party, party_type, ref_doctype = _find_confident_match(
			amount, receiving=bool(bt.deposit), company=company
		)
		if not party:
			return False

		pe = create_payment_entry_bts(
			bt_name,
			reference_number=bt.reference_number or bt_name,
			reference_date=date,
			posting_date=date,
			party_type=party_type,
			party=party,
			allow_edit=True,
		)
		pe.append("references", {
			"reference_doctype": ref_doctype,
			"reference_name": doc_name,
			"total_amount": amount,
			"outstanding_amount": amount,
			"allocated_amount": amount,
		})
		pe.insert(ignore_permissions=True)
		pe.submit()

		# Link the Bank Transaction itself to this Payment Entry — without this,
		# the PE exists and is correctly allocated, but the transaction still
		# shows Unreconciled/unallocated in the native Bank Reconciliation Tool.
		reconcile_vouchers(bt_name, json.dumps([
			{"payment_doctype": "Payment Entry", "payment_name": pe.name, "amount": amount},
		]))
		return True
	except Exception as e:
		frappe.log_error("IB Bank Auto-Match", f"{bt_name}: {e}")
		return False


@frappe.whitelist()
def import_statement(bank_account, csv_text, profile):
	"""
	Parse CSV (via the given column-mapping profile) and create Bank Transaction
	records. Skips rows where (bank_account + date + reference_number + deposit +
	withdrawal) already exists to prevent duplicates. For each new transaction,
	attempts a confident auto-match to an open Sales/Purchase document and, if
	found, creates + submits + reconciles a Payment Entry against it.
	"""
	frappe.has_permission("Bank Transaction", "create", throw=True)
	if isinstance(profile, str):
		profile = frappe.parse_json(profile)

	company = frappe.db.get_value("Bank Account", bank_account, "company")
	if not company:
		frappe.throw(_("Bank Account {0} not found or has no company.").format(bank_account))

	rows = _parse_csv_with_profile(csv_text, profile)
	created = 0
	skipped = 0
	auto_matched = 0
	errors = []

	for r in rows:
		filters = {
			"bank_account": bank_account,
			"date":         r["date"],
			"deposit":      r["deposit"],
			"withdrawal":   r["withdrawal"],
		}
		if r["reference_number"]:
			filters["reference_number"] = r["reference_number"]

		if frappe.db.exists("Bank Transaction", filters):
			skipped += 1
			continue

		try:
			bt = frappe.new_doc("Bank Transaction")
			bt.bank_account      = bank_account
			bt.company           = company
			bt.date              = r["date"]
			bt.description       = r["description"]
			bt.reference_number  = r["reference_number"]
			bt.deposit           = r["deposit"]
			bt.withdrawal        = r["withdrawal"]
			bt.currency          = frappe.db.get_value("Company", company, "default_currency") or "INR"
			bt.status            = "Pending"
			bt.insert(ignore_permissions=True)
			bt.submit()
			created += 1
			if _try_auto_match(bt.name, r["date"], company=company):
				auto_matched += 1
		except Exception as e:
			errors.append({"row": r, "error": str(e)})

	frappe.db.commit()
	return {"created": created, "skipped": skipped, "auto_matched": auto_matched, "errors": errors}

File: instabiz/instabiz/page/ib_bank_statement_import/ib_bank_statement_import.js
frappe.pages["ib-bank-statement-import"].on_page_load = function (wrapper) {
	const page = frappe.ui.make_app_page({
		parent: wrapper,
		title: "Bank Statement Import",
		single_column: true,
	});
	wrapper.page_obj = new IBBankStatementImport(page, wrapper);
};

class IBBankStatementImport {
	constructor(page, wrapper) {
		this.page = page;
		this.wrapper = wrapper;
		this.$main = $(wrapper).find(".layout-main-section");
		this._csv_text = null;
		this._preview_rows = [];
		this._bank_account = null;
		this._profile = null;
		this._render();
	}

	_render() {
		this.$main.empty().append(`
			<div class="ib-bsi-wrap">
				<div class="ib-bsi-card ib-card" style="max-width:820px;margin:0 auto;padding:28px 32px;">

					<!-- Bank account selector -->
					<div class="ib-bsi-field" style="margin-bottom:20px;">
						<label class="control-label">Bank Account</label>
						<div class="ib-bsi-bank-ctrl"></div>
					</div>

					<!-- Drop zone -->
					<div class="ib-bsi-drop" id="ib-bsi-drop">
						<svg width="40" height="40" fill="none" stroke="#adb5bd" stroke-width="1.5" viewBox="0 0 24 24">
							<path d="M4 16v2a2 2 0 002 2h12a2 2 0 002-2v-2M16 12l-4-4-4 4M12 8v8"/>
						</svg>
						<p style="margin:12px 0 4px;font-weight:600;color:var(--text-color);">
							Drop any bank's CSV here or <label for="ib-bsi-file" style="color:var(--ib-primary);cursor:pointer;text-decoration:underline;">browse</label>
						</p>
						<p style="font-size:12px;color:var(--text-muted);">Works with any bank's export — you'll map the columns once per bank account</p>
						<input type="file" id="ib-bsi-file" accept=".csv,.txt" style="display:none;">
					</div>

					<!-- Column mapping -->
					<div class="ib-bsi-mapping" style="display:none;margin-top:24px;">
						<h5 style="margin-bottom:4px;">Map Columns</h5>
						<p style="font-size:12px;color:var(--text-muted);margin-bottom:16px;">This bank account has no saved mapping yet — tell us which column is which.</p>
						<div class="ib-bsi-map-form"></div>
						<div style="margin-top:14px;">
							<label style="font-size:13px;font-weight:normal;">
								<input type="checkbox" class="ib-bsi-save-profile" checked> Remember this mapping for this bank account
							</label>
						</div>
						<div style="margin-top:16px;">
							<button class="btn btn-primary btn-sm ib-bsi-map-continue">Continue</button>
						</div>
					</div>

					<!-- Preview -->
					<div class="ib-bsi-preview" style="display:none;margin-top:24px;">
						<div class="ib-bsi-summary" style="display:flex;gap:24px;margin-bottom:16px;flex-wrap:wrap;"></div>
						<div style="overflow-x:auto;">
							<table class="table table-bordered table-sm ib-bsi-table" style="font-size:12px;margin:0;">
								<thead>
									<tr style="background:var(--subtle-accent);">
										<th>Date</th>
										<th>Description</th>
										<th>Ref No</th>
										<th style="text-align:right;">Withdrawal (Dr)</th>
										<th style="text-align:right;">Deposit (Cr)</th>
									</tr>
								</thead>
								<tbody></tbody>
							</table>
						</div>
						<div class="ib-bsi-actions" style="margin-top:20px;display:flex;gap:12px;align-items:center;">
							<button class="btn btn-primary ib-bsi-import-btn">Import</button>
							<button class="btn btn-default ib-bsi-reset-btn">Clear</button>
							<span class="ib-bsi-status" style="font-size:13px;color:var(--text-muted);"></span>
						</div>
					</div>

					<!-- Result -->
					<div class="ib-bsi-result" style="display:none;margin-top:20px;"></div>
				</div>
			</div>
		`);

		this._setup_bank_ctrl();
		this._bind();
	}

	_setup_bank_ctrl() {
		this._bank_ctrl = frappe.ui.form.make_control({
			parent: this.$main.find(".ib-bsi-bank-ctrl")[0],
			df: {
				fieldtype: "Link",
				options: "Bank Account",
				fieldname: "bank_account",
				placeholder: "Select bank account…",
				get_query: () => ({ filters: { is_company_account: 1 } }),
			},
			render_input: true,
		});
		this._bank_ctrl.refresh();
	}

	_bind() {
		const $drop = this.$main.find("#ib-bsi-drop");
		const $file = this.$main.find("#ib-bsi-file");

		$drop.on("dragover", (e) => {
			e.preventDefault();
			$drop.addClass("ib-bsi-drop--active");
		});
		$drop.on("dragleave drop", (e) => {
			e.preventDefault();
			$drop.removeClass("ib-bsi-drop--active");
			if (e.type === "drop") this._load_file(e.originalEvent.dataTransfer.files[0]);
		});
		$file.on("change", (e) => {
			if (e.target.files[0]) this._load_file(e.target.files[0]);
		});

		this.$main.on("click", ".ib-bsi-map-continue", () => this._on_mapping_continue());
		this.$main.on("click", ".ib-bsi-import-btn", () => this._do_import());
		this.$main.on("click", ".ib-bsi-reset-btn", () => this._reset());
	}

	_load_file(file) {
		if (!file) return;
		const bank_account = this._bank_ctrl.get_value();
		if (!bank_account) {
			frappe.msgprint("Select a bank account first.");
			return;
		}
		this._bank_account = bank_account;

		const reader = new FileReader();
		reader.onload = (e) => {
			this._csv_text = e.target.result;
			this._resolve_mapping();
		};
		reader.readAsText(file);
	}

	_resolve_mapping() {
		frappe.call({
			method: "instabiz.instabiz.page.ib_bank_statement_import.ib_bank_statement_import.get_saved_profile",
			args: { bank_account: this._bank_account },
			callback: (r) => {
				if (r.message) {
					this._profile = r.message;
					this._preview();
				} else {
					this._show_mapping_form();
				}
			},
		});
	}

	_show_mapping_form() {
		frappe.call({
			method: "instabiz.instabiz.page.ib_bank_statement_import.ib_bank_statement_import.get_csv_headers",
			args: { csv_text: this._csv_text },
			freeze: true,
			freeze_message: "Reading file…",
			callback: (r) => {
				if (r.exc) return;
				this._render_mapping_form(r.message.headers, r.message.samples);
			},
		});
	}

	_render_mapping_form(headers, samples) {
		const opts = ["", ...headers].map((h) => `<option value="${frappe.utils.escape_html(h)}">${frappe.utils.escape_html(h || "— none —")}</option>`).join("");
		const $form = this.$main.find(".ib-bsi-map-form").empty();

		const row = (label, cls, required) => `
			<div style="display:flex;align-items:center;gap:10px;margin-bottom:8px;">
				<label style="width:180px;font-size:13px;margin:0;">${label}${required ? " *" : ""}</label>
				<select class="form-control input-sm ${cls}" style="max-width:260px;">${opts}</select>
			</div>
		`;

		$form.append(`
			<div style="display:flex;align-items:center;gap:10px;margin-bottom:14px;">
				<label style="width:180px;font-size:13px;margin:0;">Column Layout *</label>
				<select class="form-control input-sm ib-bsi-map-mode" style="max-width:260px;">
					<option value="Separate Debit/Credit Columns">Separate Debit/Credit columns</option>
					<option value="Single Amount Column">Single Amount column</option>
				</select>
			</div>
		`);
		$form.append(row("Date Column", "ib-bsi-map-date", true));
		$form.append(row("Description Column", "ib-bsi-map-desc", false));
		$form.append(row("Reference No. Column", "ib-bsi-map-ref", false));
		$form.append(`<div class="ib-bsi-map-dc">${row("Debit/Withdrawal Column", "ib-bsi-map-debit", false)}${row("Credit/Deposit Column", "ib-bsi-map-credit", false)}</div>`);
		$form.append(`<div class="ib-bsi-map-single" style="display:none;">${row("Amount Column", "ib-bsi-map-amount", false)}${row("Type Column (Dr/Cr)", "ib-bsi-map-type", false)}`
			+ `<div style="display:flex;align-items:center;gap:10px;margin-bottom:8px;"><label style="width:180px;font-size:13px;margin:0;">Credit Indicator Value</label>`
			+ `<input type="text" class="form-control input-sm ib-bsi-map-credit-indicator" style="max-width:260px;" placeholder="e.g. CR"></div></div>`);

		if (samples && samples.length) {
			$form.append(`<p style="font-size:11px;color:var(--text-muted);margin-top:6px;">Sample row: ${frappe.utils.escape_html(JSON.stringify(samples[0]))}</p>`);
		}

		$form.find(".ib-bsi-map-mode").on("change", (e) => {
			const single = e.target.value === "Single Amount Column";
			$form.find(".ib-bsi-map-dc").toggle(!single);
			$form.find(".ib-bsi-map-single").toggle(single);
		});

		this.$main.find(".ib-bsi-mapping").show();
		this.$main.find(".ib-bsi-preview").hide();
		this.$main.find(".ib-bsi-result").hide().empty();
	}

	_on_mapping_continue() {
		const $form = this.$main.find(".ib-bsi-map-form");
		const mode = $form.find(".ib-bsi-map-mode").val();
		const col_date = $form.find(".ib-bsi-map-date").val();
		if (!col_date) {
			frappe.msgprint("Date column is required.");
			return;
		}
		const profile = {
			mode,
			col_date,
			col_description: $form.find(".ib-bsi-map-desc").val() || null,
			col_reference: $form.find(".ib-bsi-map-ref").val() || null,
			col_debit: $form.find(".ib-bsi-map-debit").val() || null,
			col_credit: $form.find(".ib-bsi-map-credit").val() || null,
			col_amount: $form.find(".ib-bsi-map-amount").val() || null,
			col_type: $form.find(".ib-bsi-map-type").val() || null,
			credit_indicator: $form.find(".ib-bsi-map-credit-indicator").val() || null,
		};
		this._profile = profile;

		const save = this.$main.find(".ib-bsi-save-profile").is(":checked");
		if (save) {
			frappe.call({
				method: "instabiz.instabiz.page.ib_bank_statement_import.ib_bank_statement_import.save_profile",
				args: { bank_account: this._bank_account, profile },
			});
		}

		this.$main.find(".ib-bsi-mapping").hide();
		this._preview();
	}

	_preview() {
		frappe.call({
			method: "instabiz.instabiz.page.ib_bank_statement_import.ib_bank_statement_import.preview_statement",
			args: { bank_account: this._bank_account, csv_text: this._csv_text, profile: this._profile },
			freeze: true,
			freeze_message: "Parsing statement…",
			callback: (r) => {
				if (r.exc) return;
				const d = r.message;
				this._preview_rows = d.rows;
				this._render_preview(d);
			},
		});
	}

	_render_preview(d) {
		const fmt = (n) => frappe.format(n, { fieldtype: "Currency" });

		const $sum = this.$main.find(".ib-bsi-summary").empty();
		[
			["Transactions", d.count, "#2d6a4f"],
			["Total Deposits", fmt(d.total_deposit), "#1a6b3c"],
			["Total Withdrawals", fmt(d.total_withdrawal), "#c0392b"],
		].forEach(([label, val, color]) => {
			$sum.append(`
				<div style="background:var(--card-bg);border:1px solid var(--border-color);border-radius:8px;padding:10px 18px;">
					<div style="font-size:11px;color:var(--text-muted);text-transform:uppercase;letter-spacing:.5px;">${label}</div>
					<div style="font-size:18px;font-weight:700;color:${color};">${val}</div>
				</div>
			`);
		});

		const $tbody = this.$main.find(".ib-bsi-table tbody").empty();
		d.rows.forEach((r) => {
			const desc = frappe.utils.escape_html(r.description || "");
			const ref  = frappe.utils.escape_html(r.reference_number || "—");
			$tbody.append(`
				<tr>
					<td style="white-space:nowrap;">${frappe.format(r.date, { fieldtype: "Date" })}</td>
					<td style="max-width:240px;overflow:hidden;text-overflow:ellipsis;white-space:nowrap;" title="${desc}">${desc}</td>
					<td>${ref}</td>
					<td style="text-align:right;color:#c0392b;">${r.withdrawal ? fmt(r.withdrawal) : "—"}</td>
					<td style="text-align:right;color:#1a6b3c;">${r.deposit ? fmt(r.deposit) : "—"}</td>
				</tr>
			`);
		});

		this.$main.find(".ib-bsi-preview").show();
		this.$main.find(".ib-bsi-result").hide().empty();
		this.$main.find(".ib-bsi-status").text("");
	}

	_do_import() {
		if (!this._preview_rows.length) return;

		this.$main.find(".ib-bsi-import-btn").prop("disabled", true).text("Importing…");

		frappe.call({
			method: "instabiz.instabiz.page.ib_bank_statement_import.ib_bank_statement_import.import_statement",
			args: { bank_account: this._bank_account, csv_text: this._csv_text, profile: this._profile },
			freeze: true,
			freeze_message: "Creating Bank Transactions…",
			callback: (r) => {
				this.$main.find(".ib-bsi-import-btn").prop("disabled", false).text("Import");
				if (r.exc) return;
				this._render_result(r.message);
			},
		});
	}

	_render_result(res) {
		const has_errors = res.errors && res.errors.length;
		const $result = this.$main.find(".ib-bsi-result").empty().show();

		let html = `
			<div class="alert alert-${res.created ? "success" : "warning"}" style="border-radius:8px;">
				<strong>${res.created} transaction${res.created !== 1 ? "s" : ""} imported.</strong>
				${res.auto_matched ? `&nbsp; ${res.auto_matched} auto-matched to an open document and reconciled with a Payment Entry.` : ""}
				${res.skipped ? `&nbsp; ${res.skipped} skipped (duplicates).` : ""}
				${has_errors ? `&nbsp; <span class="text-danger">${res.errors.length} errors.</span>` : ""}
			</div>
		`;

		if (has_errors) {
			html += `<div style="font-size:12px;color:var(--text-muted);margin-top:8px;">`;
			res.errors.forEach((e) => {
				html += `<div>• ${e.row.date} ${e.row.description}: <span class="text-danger">${e.error}</span></div>`;
			});
			html += `</div>`;
		}

		if (res.created) {
			html += `
				<div style="margin-top:16px;display:flex;gap:10px;">
					<a href="/app/bank-reconciliation-tool" class="btn btn-sm btn-default">
						Open Bank Reconciliation Tool →
					</a>
					<a href="/app/query-report/IB Bank Reconciliation" class="btn btn-sm btn-default">
						View Reconciliation Status →
					</a>
				</div>
			`;
		}

		$result.html(html);
		this.$main.find(".ib-bsi-preview").hide();
	}

	_reset() {
		this._csv_text = null;
		this._preview_rows = [];
		this._profile = null;
		this.$main.find(".ib-bsi-mapping").hide();
		this.$main.find(".ib-bsi-preview").hide();
		this.$main.find(".ib-bsi-result").hide().empty();
		this.$main.find("#ib-bsi-file").val("");
	}
}
